src/transactions/cdp.js

```diff
@@ -16,748 +30,873 @@ const MINRATIO = 140;
+  let txn1 = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
+    from: info.address,
+    suggestedParams: params,
+    to: cdp.address,
+    amount: fundingAmount, // XXX: We should code this as a const somplace / double check this
+  });
+
+  params.fee = 0;
+  let txn2 = algosdk.makeApplicationOptInTxnFromObject({
+    from: cdp.address,
+    suggestedParams: params,
+    appIndex: validatorID,
+  });
+  let r1_txns = [txn1, txn2];
+  let txn3;
+  if (!optedInGard) {
+    params.fee = 1000;
+    txn3 = createOptInTxn(params, info, gardID);
+    r1_txns.push(txn3);
+  }
+  let txn5;
+  if (!optedInGain) {
+    params.fee = 1000;
+    txn5 = createOptInTxn(params, info, gainID);
+    r1_txns.push(txn5);
+  }
+  let txn6;
+  if (VERSION != "TESTNET1" && !optedInGardian) {
+    params.fee = 1000;
+    txn6 = createOptInTxn(params, info, gardianID);
+    r1_txns.push(txn6);
+  }
+  algosdk.assignGroupID(r1_txns);
+
+  // txn 2
+  let lsig = algosdk.makeLogicSig(cdp.logic, [algosdk.encodeUint64(4)]);
+  let stxn2 = algosdk.signLogicSigTransactionObject(txn2, lsig);
+
+  // Part 2: Actually issuing the $
+
+  let collateral = openingMicroALGOs;
+  if (VERSION != "TESTNET1") {
+    collateral -= fundingAmount;
+  }
+
+  params = await getParams(0);
+  // This time could be delayed up to 20 seconds
+  let start_time = (await getCurrentUnix()) + 30;
+  txn1 = algosdk.makeApplicationCallTxnFromObject({
+    from: info.address,
+    appIndex: validatorID,
+    onComplete: 0,
+    appArgs: [enc.encode("NewPosition"), algosdk.encodeUint64(start_time)],
+    accounts: [cdp.address],
+    foreignApps: [oracleID, openFeeID],
+    foreignAssets: [gardID, accountID],
+    suggestedParams: params,
+  });
+  params.fee = 4000;
+  // console.log(typeof openingMicroALGOs);
+  txn2 = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
+    from: info.address,
+    to: cdp.address,
+    amount: collateral,
+    suggestedParams: params,
+  });
+  params.fee = 0;
+  txn3 = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
+    from: info.address,
+    to: treasury.address,
+    amount: devFees,
+    suggestedParams: params,
+  });
+  let txn4 = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
+    from: reserve.address,
+    to: info.address,
+    amount: microOpeningGard,
+    suggestedParams: params,
+    assetIndex: gardID,
+  });
+  let r2_txns = [txn1, txn2, txn3, txn4];
+  algosdk.assignGroupID(r2_txns);
+  // r2_txns.splice(3, 1)
+  let txns = r1_txns.concat(r2_txns);
+  let stxns = await signGroup(info, txns);
+  console.log(stxns);
+
+  // TODO: Adjust indexing based on including all txns
+
+  let r1_stxns = [stxns[0].blob, stxn2.blob];
+  // txn3
+  let start = 0;
+  if (!optedInGard) {
+    r1_stxns.push(stxns[2].blob);
+    start = 1;
+  }
+  if (!optedInGain) {
+    r1_stxns.push(stxns[3].blob);
+    start = 2;
+  }
+  if (VERSION != "TESTNET1" && !optedInGardian) {
+    r1_stxns.push(stxns[4].blob);
+    start = 3;
+  }
```

Comment:
There's still an issue here for users who have opted-in to GARD and GAIN, but not GARDIAN